Add some checks when trying to join the same pool already joined

emlowe opened the pull request:
Currently, if you are farming to a pool, and you join that same pool again, those transactions are merrily carried out on the chain even though this changes nothing for your farming.
Previous version would throw a cryptic error which was also a problem

This PR checks before creating any spends if the join pool URL is the same as the current pool URL
Fixes #7292

emlowe committed:
Add some checks when trying to join the same pool already joined

emlowe committed:
Add test to test the specific cli case

emlowe committed:
Merge remote-tracking branch 'origin/main' into EL.pool-join-checks

emlowe committed:
Remove line added during debugging

emlowe committed:
Add test

emlowe committed:
misc updates

emlowe committed:
remove cli test

emlowe committed:
Merge remote-tracking branch 'origin/main' into EL.pool-join-checks

emlowe committed:
final tests for joining same pool again

altendky commented on chia/_tests/pools/test_pool_rpc.py:
```diff
@@ -1020,3 +1022,48 @@ async def status_is_leaving_no_blocks() -> bool:
+        await time_out_assert(45, status_is_farming_to_pool, True, wallet_id)
+
+        # Test joining the same pool via the RPC client
+        with pytest.raises(ValueError, match="Already farming to pool"):
```
Needs an import as well.

```suggestion
        with pytest.raises(ResponseFailureError, match="Already farming to pool"):
```

Quexington commented on chia/_tests/pools/test_pool_rpc.py:
```diff
@@ -1020,3 +1022,48 @@ async def status_is_leaving_no_blocks() -> bool:
         # Eventually, leaves pool
         assert await status_is_farming_to_pool()
+
+    @pytest.mark.anyio
```
Is there a reason this isn't `WalletTestFramework`?

emlowe replied:
nothing in this file was using WalletTestFramework at the time, so I followed the conventions here. But I think your question is a good one and I'll look to convert it to use the newer concepts

emlowe replied:
Updated test to use WTF

emlowe committed:
Merge remote-tracking branch 'origin/main' into EL.pool-join-checks

emlowe committed:
Update check from code review

emlowe committed:
update test framework

altendky commented on chia/cmds/plotnft_funcs.py:
```diff
@@ -307,6 +311,13 @@ async def join_pool(
+        pool_wallet_info.current.state == PoolSingletonState.FARMING_TO_POOL.value
+        and pool_wallet_info.current.pool_url == pool_url
+    ):
+        raise CliRpcConnectionError(f"Wallet id: {wallet_id} is already farming to pool {pool_url}")
```
`CliRpcConnectionError` doesn't seem right here.

https://github.com/Chia-Network/chia-blockchain/blob/30c4ef2689cf69578a936e7338fc1d089977d9d4/chia/util/errors.py#L348-L353

emlowe committed:
Merge remote-tracking branch 'origin/main' into EL.pool-join-checks

emlowe committed:
Use click.ClickException for error

arvidn reviewed (dismissed):
Is there already a check to see if the wallet is in sync or not? I imagine if we're not in sync the `current` pool target may not actually be current.

emlowe commented:
> Is there already a check to see if the wallet is in sync or not? I imagine if we're not in sync the `current` pool target may not actually be current.

Interesting question - if your syncing your wallet, and try to join a pool, you may get this error early because this error would come first (I think). So that may possibly be confusing to the end user. I'll have to revisit to make sure

emlowe committed:
Add check for wallet synced up front as the first thing checked for joining a pool

emlowe commented:
The RPC had a check for the wallet synced - so I moved the farm pool check below the sync check.
The front-end didn't have this check at all, so I added it and a test.

I'll note there is some duplication here about where to check for errors. The RPC obviously has to check all the items, but when the CLI front-end also checks them we get code duplication that we should work probably work to avoid. But the RPC errors are often handled rather poorly from a UX perspective in our code and returning a `ClickException` is nice up front in the CLI.

emlowe committed:
Add test for unsynced RPC call

emlowe commented:
My aversion to coverage errors made me write a quick test to cover the unsynced case for the RPC too - even though that code was already present, but apparently untested.

Quexington commented:
> My aversion to coverage errors made me write a quick test to cover the unsynced case for the RPC too - even though that code was already present, but apparently untested.

Just so it's documented, there's a trade-off here between test coverage and keeping PRs focused.  Doesn't really matter in this instance but not sure it's always correct.
